### PR TITLE
Fix utxo sync delay.

### PR DESCRIPTION
**Pull request by raduom**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

**Review by koslambrou — APPROVED** [review]

File: plutus-chain-index/app/Marconi/Index/Utxo.hs [hunk 1]
```diff
@@ -98,13 +100,10 @@ open
   -> Depth
   -> IO UtxoIndex
 open dbPath (Depth k) = do
-  ix <- fromJust <$> Ix.newBoxed query store onInsert k ((k + 1) * 4) dbPath
+  ix <- fromJust <$> Ix.newBoxed query store onInsert k ((k + 1) * 2) dbPath
```
Comment (andreabedini) [inline plutus-chain-index/app/Marconi/Index/Utxo.hs:103]: Can you leave a note explaining this `(k + 1) * 2`?

File: plutus-chain-index/app/Marconi/Index/Utxo.hs [hunk 2]
```diff
@@ -114,7 +113,14 @@ query
   -> IO Result
 query ix addr updates = do
   -- SELECT all utxos that have not been spent.
-  storedUtxos <- SQL.query (ix ^. Ix.handle) "SELECT address, txId, inputIx FROM utxos LEFT JOIN spent ON utxos.txId = spent.txId AND utxos.inputIx = spent.inputIx WHERE utxos.txId IS NULL AND utxos.address = ?" (Only addr)
+  let c = ix ^. Ix.handle
+  -- Create indexes initially. When created this should be a no-op.
```
Comment (andreabedini) [inline plutus-chain-index/app/Marconi/Index/Utxo.hs:117]: Doing it a query time is a clever idea. The user can decide how much it wants to wait before creating the indices. 

How is  SQLite going to handle creating the indexes while inserting new data? is it going to block the writes or can it do it concurrently?

**Review by andreabedini — COMMENTED** [review]
Good stuff, I left some comments.

File: plutus-chain-index/app/Marconi/Index/Utxo.hs [hunk 3]
```diff
@@ -136,10 +142,14 @@ store ix = do
     SQL.execute c "INSERT INTO utxos (address, txId, inputIx) VALUES (?, ?, ?)"
   forM_ spent $
     SQL.execute c "INSERT INTO spent (txId, inputIx) VALUES (?, ?)"
-
-  SQL.execute_ c "DELETE FROM utxos WHERE utxos.rowid IN (SELECT utxos.rowid FROM utxos LEFT JOIN spent on utxos.txId = spent.txId AND utxos.inputIx = spent.inputIx WHERE spent.txId IS NOT NULL)"
   SQL.execute_ c "COMMIT"
 
+  rndCheck <- createSystemRandom >>= uniformR (1 :: Int, 200)
+  when (rndCheck == 42) $ do
```
[the remark this answers is not in the document]
Comment (raduom) [reply plutus-chain-index/app/Marconi/Index/Utxo.hs:148]: Every 200 * 2 (buffer) * 2160 (K) blocks. I've changed it to 100 so it collects more often.

File: plutus-chain-index/app/Marconi/Index/Utxo.hs (hunk 1 above)
Comment (raduom) [reply plutus-chain-index/app/Marconi/Index/Utxo.hs:103]: That argument represents the buffer used in the indexer. I am simply running the indexer and trying to adjust the arguments to get a reasonable memory usage / time compromise.

File: plutus-chain-index/app/Marconi/Index/Utxo.hs (hunk 2 above)
Comment (raduom) [reply plutus-chain-index/app/Marconi/Index/Utxo.hs:117]: I think it does the right thing (tm), but I have not tested it.

File: plutus-chain-index/plutus-chain-index.cabal [hunk 4]
```diff
@@ -146,6 +146,7 @@ executable marconi
     , bytestring
     , containers
     , lens
+    , mwc-random
```
[the remark this answers is not in the document]
Comment (raduom) [reply plutus-chain-index/plutus-chain-index.cabal:149]: I did try to use the random package initially, but `mwc-random` has a better interface imho. It's also what we use in `plutus-pab-executables`.

[the remark this answers is not in the document]
Comment (andreabedini) [reply plutus-chain-index/plutus-chain-index.cabal:149]: ok, no problem if we have a precedent in using `mwc-random`

File: plutus-chain-index/app/Marconi/Index/Utxo.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreabedini) [reply plutus-chain-index/app/Marconi/Index/Utxo.hs:148]: Can you leave a note here too?

**Review by andreabedini — APPROVED** [review]
Awesome

**Pull request merged**